lib/teleterm/daemon/daemon.go

```diff
@@ -215,55 +184,98 @@ func (s *Service) RemoveGateway(ctx context.Context, gatewayURI string) error {
+	//
+	// This is preferable to removing the gateway from the hash map. Since Connect remembers the port
+	// used for a particular db server + username pair, the user would be able to reopen the same
+	// gateway on the occupied port only to see an error.
 	if err := gateway.Close(); err != nil {
 		return trace.Wrap(err)
 	}
 
-	// remove closed gateway from list
-	for index := range s.gateways {
-		if s.gateways[index] == gateway {
-			s.gateways = append(s.gateways[:index], s.gateways[index+1:]...)
-			return nil
-		}
-	}
+	delete(s.gateways, gateway.URI.String())
 
-	return trace.NotFound("gateway %v not found in gateway list", gateway.URI.String())
+	return nil
 }
 
 // RestartGateway stops a gateway and starts a new one with identical parameters.
 // It also keeps the original URI so that from the perspective of Connect it's still the same
 // gateway but with fresh certs.
 func (s *Service) RestartGateway(ctx context.Context, gatewayURI string) error {
-	gateway, err := s.FindGateway(gatewayURI)
+	oldGateway, err := s.FindGateway(gatewayURI)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	if err := s.removeGateway(gateway); err != nil {
+	if err := s.removeGateway(oldGateway); err != nil {
 		return trace.Wrap(err)
 	}
 
 	newGateway, err := s.createGateway(ctx, CreateGatewayParams{
-		TargetURI:             gateway.TargetURI,
-		TargetUser:            gateway.TargetUser,
-		TargetSubresourceName: gateway.TargetSubresourceName,
-		LocalPort:             gateway.LocalPort,
+		TargetURI:             oldGateway.TargetURI,
+		TargetUser:            oldGateway.TargetUser,
+		TargetSubresourceName: oldGateway.TargetSubresourceName,
+		LocalPort:             oldGateway.LocalPort,
 	})
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	newGateway.URI = gateway.URI
+	// s.createGateway adds a gateway under a random URI, so we need to place the new gateway under
+	// the URI of the old gateway.
+	delete(s.gateways, newGateway.URI.String())
+	newGateway.URI = oldGateway.URI
+	s.gateways[oldGateway.URI.String()] = newGateway
 
 	return nil
 }
 
+// FindGateway finds a gateway by URI
+func (s *Service) FindGateway(gatewayURI string) (*gateway.Gateway, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	gateway, err := s.findGateway(gatewayURI)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return gateway, nil
+}
+
+// findGateway assumes that mu is already held by a public method.
+func (s *Service) findGateway(gatewayURI string) (*gateway.Gateway, error) {
+	if gateway, ok := s.gateways[gatewayURI]; ok {
+		return gateway, nil
+	}
+
+	return nil, trace.NotFound("gateway is not found: %v", gatewayURI)
+}
+
+// ListGateways lists gateways
+func (s *Service) ListGateways() []*gateway.Gateway {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	gws := make([]*gateway.Gateway, 0, len(s.gateways))
+	for _, gateway := range s.gateways {
+		gateway := gateway
```

Comment:
Should you return pointers to the gateways in the list? Does that allow the caller to unsafely modify the gateways out from under you?

Reply:
Right, I should not return pointers here. The same can be said about `FindGateway`.

I'll remove `FindGateway` altogether so that only `findGateway` exists. `findGateway` assumes you hold a lock for `s.gateways` so this will help with fixing the lock issues in `RemoveGateway` and `RestartGateway`.